relayer/message_relayer.go

```diff
@@ -548,27 +586,176 @@ func (r *messageRelayer) aggregateSignatures(signatureMap map[int]blsSignatureBu
 	return aggSig, vdrBitSet, nil
 }
 
+//
+// Database access
+//
+
+// Determines the height to process from. There are three cases:
+// 1) The database contains the latest processed block data for the chain
+//   - In this case, we return the maximum of the latest processed block and the configured processHistoricalBlocksFromHeight
+//
+// 2) The database has been configured for the chain, but does not contain the latest processed block data
+//   - In this case, we return the configured processHistoricalBlocksFromHeight
+//
+// 3) The database does not contain the any information for the chain.
+//   - In this case, we return the configured processHistoricalBlocksFromHeight if it is set, otherwise
+//     we return the chain head.
+func (r *messageRelayer) calculateStartingBlockHeight(processHistoricalBlocksFromHeight uint64) (uint64, error) {
+	latestProcessedBlock, err := r.getLatestProcessedBlockHeight()
+	if database.IsKeyNotFoundError(err) {
+		// The database does not contain the latest processed block data for the chain,
+		// use the configured process-historical-blocks-from-height instead.
+		// If process-historical-blocks-from-height was not configured, start from the chain head.
+		if processHistoricalBlocksFromHeight == 0 {
+			return r.setProcessedBlockHeightToLatest()
+		}
+		return processHistoricalBlocksFromHeight, nil
+	} else if err != nil {
+		// Otherwise, we've encountered an unknown database error
+		r.logger.Error(
+			"failed to get latest block from database",
+			zap.String("relayerKey", r.relayerKey.CalculateRelayerKey().String()),
+			zap.Error(err),
+		)
+		return 0, err
+	}
```

Comment:
Just to confirm my understanding: We process/relay blocks in order. As a result, if we have processed the latest block height, then we know that there are no blocks prior to the latest block height that we have not processed. Is this correct?

Reply:
That's not necessarily the case when the relayer is processing historical blocks. We process historical messages and live incoming messages concurrently, so we can't make assumptions about ordering until we're done catching up. Live incoming messages are always processed in order relative to each other.

These ordering assumptions can be tricky to reason about. We have some ideas on how to relax those ordering requirements for both historical and live incoming messges. See the section `Relax Ordering Requirement` in https://github.com/ava-labs/awm-relayer/issues/234